terminal-emulator/src/main/java/com/termux/terminal/TerminalEmulator.java

```diff
@@ -0,0 +1,2447 @@
+                        startRow = justRow ? mCursorRow : 0;
+                        endCol = mCursorCol + 1;
+                        endRow = mCursorRow + 1;
+                        break;
+                    case 2: // Erase all of the display/line.
+                        startCol = 0;
+                        startRow = justRow ? mCursorRow : 0;
+                        endCol = mColumns;
+                        endRow = justRow ? (mCursorRow + 1) : mRows;
+                        break;
+                    default:
+                        unknownSequence(b);
+                        break;
+                }
+                long style = getStyle();
+                for (int row = startRow; row < endRow; row++) {
+                    for (int col = startCol; col < endCol; col++) {
+                        if ((TextStyle.decodeEffect(mScreen.getStyleAt(row, col)) & TextStyle.CHARACTER_ATTRIBUTE_PROTECTED) == 0)
+                            mScreen.setChar(col, row, fillChar, style);
+                    }
+                }
+                break;
+            case 'h':
+            case 'l':
+                if (mArgIndex >= mArgs.length) mArgIndex = mArgs.length - 1;
+                for (int i = 0; i <= mArgIndex; i++)
+                    doDecSetOrReset(b == 'h', mArgs[i]);
+                break;
+            case 'n': // Device Status Report (DSR, DEC-specific).
+                switch (getArg0(-1)) {
+                    case 6:
+                        // Extended Cursor Position (DECXCPR - http://www.vt100.net/docs/vt510-rm/DECXCPR). Page=1.
+                        mSession.write(String.format(Locale.US, "\033[?%d;%d;1R", mCursorRow + 1, mCursorCol + 1));
+                        break;
+                    default:
+                        finishSequence();
+                        return;
+                }
+                break;
+            case 'r':
+            case 's':
+                if (mArgIndex >= mArgs.length) mArgIndex = mArgs.length - 1;
+                for (int i = 0; i <= mArgIndex; i++) {
+                    int externalBit = mArgs[i];
+                    int internalBit = mapDecSetBitToInternalBit(externalBit);
+                    if (internalBit == -1) {
+                        mClient.logWarn(LOG_TAG, "Ignoring request to save/recall decset bit=" + externalBit);
+                    } else {
+                        if (b == 's') {
+                            mSavedDecSetFlags |= internalBit;
+                        } else {
+                            doDecSetOrReset((mSavedDecSetFlags & internalBit) != 0, externalBit);
+                        }
+                    }
+                }
+                break;
+            case '$':
+                continueSequence(ESC_CSI_QUESTIONMARK_ARG_DOLLAR);
+                return;
+            default:
+                parseArg(b);
+        }
+    }
+
+    public void doDecSetOrReset(boolean setting, int externalBit) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This method contains 78 lines of code, not including blank lines or lines with only comments, Java punctuation characters, identifiers, or literals. By comparison, 99% of the methods in the CodeGuru reference dataset contain fewer lines of code. Large methods might be difficult to read and have logic that is hard to understand and test. We recommend that you simplify this method or break it into multiple methods.